src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java

```diff
@@ -0,0 +1,399 @@
+            Pointer replacementPointer = replacement.getBlock().ptr;
+
+            int dominatesOld = dominates(newPointer, origPointer);
+            int dominatesNew = dominates(newPointer, replacementPointer);
+
+            if (dominatesOld == 0 || dominatesNew == 0) {
+                // can not replace node in this case since new node does not dominate one of the replacements
+                return false;
+            }
+            // new node has to dominate all uses
+            boolean usesDominated = dominatesUses(newNode, orig);
+            boolean usesDominatedRep = dominatesUses(newNode, replacement);
+
+            if (hasCallAsChild(orig) || hasCallAsChild(replacement)) {
+                System.out.println("call as child do not replace");
+                return false;
+            }
+            if (usesDominated && usesDominatedRep) {
+                //      System.out.println("replaced node " + orig + " with " + newNode + " in graph " + g);
+                Graph.exchange(orig, newNode);
+                Graph.exchange(replacement, newNode);
+                return true;
+            } else {
+                return false;
+            }
+        }
+        return false;
+    }
+
+    private boolean hasCallAsChild(Node n) {
+        for (BackEdges.Edge edge : BackEdges.getOuts(n)) {
+            Node node = edge.node;
+            if (node.getOpCode() == binding_irnode.ir_opcode.iro_Call) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private int dominates(Pointer source, Pointer target) {
+        int dominates = 0;
+
+        if (!dominateMap.containsKey(source)) {
+            // we have no entry yet, compute both values
+            dominates = binding_irdom.block_dominates(source, target);
+
+            HashMap<Pointer, Integer> cachedResults = new HashMap<>();
+            cachedResults.put(target, dominates);
+            dominateMap.put(source, cachedResults);
+
+        } else {
+            Map<Pointer, Integer> cachedResults = dominateMap.get(source);
+            dominates = cachedResults.computeIfAbsent(target, p -> binding_irdom.block_dominates(source, target));
+        }
+        return dominates;
+    }
+
+    private boolean dominatesUses(Node newNode, Node orig) {
+        for (BackEdges.Edge edge : BackEdges.getOuts(orig)) {
+            Node n = edge.node;
+            if (dominates(newNode.getBlock().ptr, n.getBlock().ptr) == 0) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+
+    private final class CSEVisitor extends NodeVisitor.Default {
+
+        private final Map<TargetValue, Node> constCache = new HashMap<>();
+        private final Map<NodePreds, Node> nodeCache = new HashMap<>();
+
+        @RequiredArgsConstructor
+        private class NodePreds {
+            private final Node[] preds;
+            private final binding_irnode.ir_opcode opcode;
+            private final Mode mode;
```

Comment:
The `Mode` is only included in your `equals`. This breaks the contract concerning the behavior of `equals` and `hashCode`. Maybe consider just letting `lombok` generate the two methods for you.